### PR TITLE
feat: Add `testing.assert_frame_equal`

### LINKED ISSUE #3129: Support `assert_frame_equal`

### PR DESCRIPTION
**Pull request by FBruzzesi**
## What type of PR is this? (check all applicable)

- [ ] 💾 Refactor
- [x] ✨ Feature
- [ ] 🐛 Bug Fix
- [ ] 🔧 Optimization
- [ ] 📝 Documentation
- [ ] ✅ Test
- [ ] 🐳 Other

## Related issues

- Closes #3129
- Natural follow up on #2983
- Possible follow up #3221

## Checklist

- [x] Code follows style guide (ruff)
- [x] Tests added
- [x] Documented the changes

## If you have comments or can explain your changes, please do so below

50%+ of the changes are docstrings and tests

### CODE REVIEW

File: tests/testing/assert_frame_equal_test.py [hunk 1]
```diff
@@ -0,0 +1,249 @@
...
+    left = nw.from_native(constructor(data)).select(b_expr)
+
+    msg = "`check_row_order=False` is not supported (yet) with only nested data type."
+    with pytest.raises(NotImplementedError, match=re.escape(msg)):
+        assert_frame_equal(left, left, check_row_order=False)
+
+
+# TODO(FBruzzesi): Do we even this? What should we check?
+# The best way could be to migrate away from `assert_equal_data` in the test suite
+def test_values_mismatch(constructor: Constructor) -> None: ...
```
Comment (FBruzzesi) [inline tests/testing/assert_frame_equal_test.py]: Looking for help in finding edge cases here 😇

File: narwhals/testing/asserts/utils.py [hunk 2]
```diff
@@ -7,23 +8,42 @@
...
+            "columns are not in the same order",
+            "dtypes do not match",
+            "height (row count) mismatch",
+            "implementation mismatch",
+        ]
+        | str
+        # NOTE: `| str` makes the literals above redundant, but they still show
+        # up when typing as autocompletion.
+        # The reason to have `str` is due to the fact that other details are dynamic
+        # and depend upon which columns lead to the assertion error.
```
Comment (FBruzzesi) [inline narwhals/testing/asserts/utils.py]: @dangotbanned do you see a better way of doing this?

Comment (dangotbanned) [reply narwhals/testing/asserts/utils.py]: My first thought would be to keep the `*Detail` aliases for static portion of the message.

I haven't looked through to see how the dynamic stuff works yet - particularly where it is sourced and where it gets inserted - but if you had these things separated into 2 parameters then it might work?

E.g. if you added another parameter with `not_thing: str = ""`, that could just get skipped when empty?

Comment (dangotbanned) [reply narwhals/testing/asserts/utils.py]: This example uses `None` and adds line breaks when you have content

https://github.com/vega/altair/blob/31c7f8a4b74f824311af459f17b8c5c65c32ba3c/altair/utils/deprecation.py

**Comment by dangotbanned** [comment]
@FBruzzesi you might be able to adapt this test from (https://github.com/vega/altair/pull/3922) into something for timezones?


At least the source data may give you something to compare 🙂 

```py
import polars as pl

start = pl.datetime(2023, 11, 5, time_zone="US/Mountain")
pl.select(
    datetime=pl.datetime_range(start, start.dt.offset_by("3h"), "1h"),
    value=pl.int_range(10, 50, 10),
)
```

```
shape: (4, 2)
┌───────────────────────────┬───────┐
│ datetime                  ┆ value │
│ ---                       ┆ ---   │
│ datetime[μs, US/Mountain] ┆ i64   │
╞═══════════════════════════╪═══════╡
│ 2023-11-05 00:00:00 MDT   ┆ 10    │
│ 2023-11-05 01:00:00 MDT   ┆ 20    │
│ 2023-11-05 01:00:00 MST   ┆ 30    │
│ 2023-11-05 02:00:00 MST   ┆ 40    │
└───────────────────────────┴───────┘
```

**Comment by FBruzzesi** [comment]
Hey @dangotbanned 👋🏼 I am missing some context for the https://github.com/narwhals-dev/narwhals/pull/3220#issuecomment-3570084015

What would you like to see achieved? If it's dataframe with one or more datetime with timezones colujmns, then those checks are delegated to the series equality 👀 (via `assert_series_equal`), which is then dispatched with the usual narwhals mechanism

```py
is_not_equal_mask = left != right
if is_not_equal_mask.any(): ...
```

**Comment by dangotbanned** [comment]
> Hey @dangotbanned 👋🏼 I am missing some context for the [#3220 (comment)](https://github.com/narwhals-dev/narwhals/pull/3220#issuecomment-3570084015)
> 
> What would you like to see achieved? 

Oof my bad, I was supposed to have replied on the (https://github.com/narwhals-dev/narwhals/pull/3220#discussion_r2442074358) thread 😅

> Looking for help in finding edge cases here 😇

Maybe I'm naive 😉, but I hadn't come across that timezone before - so I thought it might help?

**Comment by FBruzzesi** [comment]
> [!WARNING]
>  This proposal might end up be a bit overwhelming in one go

The best way to test this implementation would be to start rolling it out as proposed in #3221 - I can branch out from here and start doing that, maybe on a specific subset of tests (say `tests/frame/`)

**Comment by FBruzzesi** [comment]
@dangotbanned pinging you here once more. If you are happy with the state of this, I would be happy to go forward and merging.

As a disclaimer, one possible blocker is https://github.com/narwhals-dev/narwhals/issues/2939, but I rather have something that works in most cases for now.

I will try to merge it right after a release, so that I have some more time to explore the followup proposed in #3221

File: narwhals/testing/asserts/frame.py [hunk 3]
```diff
@@ -0,0 +1,265 @@
...
+        if left_not_in_right := sorted(lset.difference(rset)):
+            raise_frame_assertion_error(
+                detail=f"{left_not_in_right} in left, but not in right",
+                left=lset,
+                right=rset,
+            )
+        if right_not_in_left := sorted(rset.difference(lset)):  # pragma: no cover
+            # NOTE: the `pragma: no cover` flag is due to a false negative.
+            # The last test in `test_check_schema_mismatch` does cover this case.
+            raise_frame_assertion_error(
```
Comment (dangotbanned) [inline narwhals/testing/asserts/frame.py]: Was the issue for coverage the unreachable falsy branch?

It looks like if we got here, `right_not_in_left` must be non-empty.
Otherwise, how would `lset != rset` be true?

File: narwhals/testing/asserts/frame.py [hunk 4]
```diff
@@ -0,0 +1,265 @@
...
+    Implementation.POLARS,
+    Implementation.DASK,
+}
+
+
+def assert_frame_equal(
+    left: DataFrameT | LazyFrameT,
+    right: DataFrameT | LazyFrameT,
+    *,
+    check_row_order: bool = True,
```
Comment (dangotbanned) [inline narwhals/testing/asserts/frame.py:32]: Have you thought about providing a `backend` argument to forward into `LazyFrame.collect`?

File: narwhals/testing/asserts/frame.py [hunk 5]
```diff
@@ -0,0 +1,265 @@
...
+    # NOTE: Here `impl` comes from the original dataframe, not the `.collect`-ed one, and
+    # it's used to distinguish between backends that do and do not guarantee row order.
+    _check_schema_equal(
+        left, right, check_dtypes=check_dtypes, check_column_order=check_column_order
+    )
+
+    left_len, right_len = len(left), len(right)
+    if left_len != right_len:
+        raise_frame_assertion_error("height (row count) mismatch", left_len, right_len)
+    # TODO(FBruzzesi): Should we return early if row count is zero?
```
Comment (dangotbanned) [inline narwhals/testing/asserts/frame.py]: Yeah it looks like everything after this point can be skipped for 0 rows

File: narwhals/testing/asserts/frame.py [hunk 6]
```diff
@@ -0,0 +1,265 @@
...
+    # Adapted from https://github.com/pola-rs/polars/blob/afdbf3056d1228cf493901e45f536b0905cec8ea/crates/polars-testing/src/asserts/utils.rs#L829
+    # NOTE: Here `impl` comes from the original dataframe, not the `.collect`-ed one, and
+    # it's used to distinguish between backends that do and do not guarantee row order.
+    _check_schema_equal(
+        left, right, check_dtypes=check_dtypes, check_column_order=check_column_order
+    )
+
+    left_len, right_len = len(left), len(right)
+    if left_len != right_len:
+        raise_frame_assertion_error("height (row count) mismatch", left_len, right_len)
```
Comment (dangotbanned) [inline narwhals/testing/asserts/frame.py:189]: Oh this reminded me that we don't have `DataFrame.{height,width}`!

File: narwhals/testing/asserts/utils.py [hunk 7]
```diff
@@ -43,3 +55,17 @@ def raise_series_assertion_error(
...
+    left: Any,
+    right: Any,
+    *,
+    cause: Exception | None = None,
+) -> Never:
+    # NOTE: `DataFramesDetail | str` makes the literal (`DataFramesDetail`) redundant.
+    # However, the suggestions still show up as autocompletion in the editor when typing.
+    # The reason to have `str` is due to the fact that some details are dynamic
+    # and depend upon which columns lead to the assertion error.
+    raise_assertion_error("DataFrames", detail, left, right, cause=cause)
```
Comment (dangotbanned) [inline narwhals/testing/asserts/utils.py]: I did hint at a solution for this in https://github.com/narwhals-dev/narwhals/pull/3220#discussion_r2442234992

Here's an entire module of examples 😄

- https://github.com/narwhals-dev/narwhals/blob/058782d1992b2787624b1f7b27f5612d04341e91/narwhals/_plan/exceptions.py

---

But I'll use one from here to demonstrate:

https://github.com/narwhals-dev/narwhals/blob/00f074fdda364403675b65f98dbb129e4442276f/narwhals/testing/asserts/frame.py#L233-L247

This could be calls to a *different* function, where you pass the different columns separately:

```py
raise_frame_names_assertion_error(  # or a good name
    difference=left_not_in_right,
    detail="in left, but not in right",
    left=lset,
    right=rset,
)

raise_frame_names_assertion_error(
    difference=right_not_in_left
    detail="in right, but not in left",
    left=lset,
    right=rset,
)
```

File: narwhals/testing/asserts/frame.py (hunk 3 above)
Comment (FBruzzesi) [reply narwhals/testing/asserts/frame.py]: Fair enough! I liked the symmetry, but it's fine not to have it 😂

Actually removing the first condition also fixes it

File: narwhals/testing/asserts/utils.py (hunk 7 above)
Comment (FBruzzesi) [reply narwhals/testing/asserts/utils.py]: We need custom prefix and suffix: https://github.com/narwhals-dev/narwhals/pull/3220/commits/154f2fe4c1b30ee3b047073898485c358058cc5a

**Comment by FBruzzesi** [comment]
~mmmh I can't replicate the mkdocs build failure locally 🤔~


Fixed in https://github.com/narwhals-dev/narwhals/pull/3220/commits/43e6870862590a857800559420a00c471d932b1f
